core/src/parser/utils.rs

```diff
@@ -536,30 +522,88 @@ where
+        let RichTerm {
+            term: t2,
+            pos: pos2,
+        } = t2;
+        match (t1.into_owned(), t2.into_owned()) {
+            (Term::Record(rd1), Term::Record(rd2)) => {
+                let split::SplitResult {
+                    left,
+                    center,
+                    right,
+                } = split::split(rd1.fields, rd2.fields);
+                let mut fields = IndexMap::with_capacity(left.len() + center.len() + right.len());
+                fields.extend(left);
+                fields.extend(right);
+                for (id, (field1, field2)) in center.into_iter() {
+                    fields.insert(id, merge_fields(id_span, field1, field2));
+                }
+                Term::Record(RecordData::new(
+                    fields,
+                    RecordAttrs::combine(rd1.attrs, rd2.attrs),
+                    None,
+                ))
+                .into()
+            }
+            (t1, t2) => mk_term::op2(
+                BinaryOp::Merge(MergeLabel {
+                    span: id_span,
+                    kind: MergeKind::PiecewiseDef,
+                }),
+                RichTerm::new(t1, pos1),
+                RichTerm::new(t2, pos2),
+            ),
+        }
+    }
 
-    let metadata = FieldMetadata::combine(field1.metadata, field2.metadata);
+    let (value, priority) = match (field1.value, field2.value) {
+        (Some(t1), Some(t2)) if field1.metadata.priority == field2.metadata.priority => (
+            Some(merge_values(id_span, t1, t2)),
+            field1.metadata.priority,
+        ),
+        (Some(t), _) if field1.metadata.priority > field2.metadata.priority => {
+            (Some(t), field1.metadata.priority)
+        }
+        (_, Some(t)) if field1.metadata.priority < field2.metadata.priority => {
+            (Some(t), field2.metadata.priority)
+        }
+        (Some(t), None) => (Some(t), field1.metadata.priority),
+        (None, Some(t)) => (Some(t), field2.metadata.priority),
+        (None, None) => (None, Default::default()),
+        _ => unreachable!(),
+    };
 
     // At this stage, pending contracts aren't filled nor meaningful, and should all be empty.
     debug_assert!(field1.pending_contracts.is_empty() && field2.pending_contracts.is_empty());
     Field {
         value,
-        metadata,
-        pending_contracts: field1.pending_contracts,
+        // [`FieldMetadata::combine`] produces subtly different behaviour from
+        // the runtime merging code, which is what we need to replicate here
+        metadata: FieldMetadata {
+            doc: merge_doc(field1.metadata.doc, field2.metadata.doc),
+            annotation: Combine::combine(field1.metadata.annotation, field2.metadata.annotation),
+            opt: field1.metadata.opt && field2.metadata.opt,
+            not_exported: field1.metadata.not_exported || field2.metadata.not_exported,
+            priority,
+        },
+        pending_contracts: Vec::new(),
```

Comment:
It's not very important, but the idea of using `field1.pending_contracts` was to avoid dropping an empty `Vec` just to recreate one just after.

Reply:
My thought process was: creating an empty `Vec` doesn't allocate and if for some reason `field1.pending_contracts` was at some point nonempty, then got cleared and then we end up here, we might as well drop its allocation.

In the end, this probably never happens and we're talking about 16 bytes being initialized to 0. 🤷

Reply:
I think it's 24 bytes with cap (ptr + size + cap), which is huge! Jokes aside, yes, it's unimportant. And TIL that an empty vector doesn't allocate before something is pushed in it.